publish.go

```diff
@@ -2,31 +2,56 @@ package legs
 
 import (
 	"context"
-	"sync/atomic"
 
+	dt "github.com/filecoin-project/go-data-transfer"
 	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
 )
 
 type legPublisher struct {
-	transfer *LegTransport
+	topic   *pubsub.Topic
+	onClose func() error
 }
 
 // NewPublisher creates a new legs publisher
-func NewPublisher(ctx context.Context, dt *LegTransport) (LegPublisher, error) {
-
-	// Track how many publishers are using this transport
-	dt.addRefc()
+func NewPublisher(ctx context.Context,
+	host host.Host,
+	ds datastore.Batching,
+	lsys ipld.LinkSystem,
+	topic string) (LegPublisher, error) {
+	ss, err := newSimpleSetup(ctx, host, ds, lsys, topic)
+	if err != nil {
+		return nil, err
+	}
+	return &legPublisher{ss.t, ss.onClose}, nil
+}
 
-	return &legPublisher{
-		transfer: dt}, nil
+// NewPublisherFromExisting instantiates go-legs publishing on an existing
+// data transfer instance
+func NewPublisherFromExisting(ctx context.Context,
```

Comment:
the thought is that this is what we would expect a lotus to make use of, right?

Reply:
yea, and honestly the reference provider would do well to do it also